plugins/dead_code_elimination/dead_code_elimination_main.ml

```diff
@@ -23,95 +38,155 @@ end
 let computed_def_use sub =
   def_use_collector#visit_sub sub (Var.Set.empty,Var.Set.empty)
 
-let sub_args sub =
-  Term.enum arg_t sub |>
-  union ~init:Var.Set.empty ~f:(fun arg -> Exp.free_vars (Arg.rhs arg))
-let free_vars sub = Set.union (Sub.free_vars sub) (sub_args sub)
-
-let compute_dead protected sub =
+let compute_dead checked protected sub =
   let defs,uses = computed_def_use sub in
   let dead = Set.diff defs uses in
   let live v = not (Set.mem dead v) in
   (object inherit [Tid.Set.t] Term.visitor
-      method! enter_def t dead =
-        let v = Def.lhs t in
-        if Set.mem protected (Var.base v) || live v
-        then dead
-        else (Set.add dead (Term.tid t))
+    method! enter_def t dead =
+      let v = Def.lhs t in
+      if Set.mem protected (Var.base v) || live v || not (checked v)
+      then dead
+      else Set.add dead (Term.tid t)
   end)#visit_sub sub Tid.Set.empty
 
-(* a simple constant propagation, that will propagate constant
-   expressions from virtual variables that are assigned only once to
-   the point of usage (if any).
-
-   1. we propagate only virtuals since they are not used/clobbered by
-      function calls. We can't touch real registers, as we don't know
-      the calling convention. And even if we know, a malicious program
-      may easily break it.
-
-   2. instead of a full-fledges reaching definition analysis we are
-      relying on a simple over-approximation: if a variable is defined
-      only once, then it is never redifined. Enough for reaping the
-      low hanging fruits.
-
-   Note, the input is not required to be in SSA.
- *)
-let propagate_consts sub =
-  let vars = (object inherit [exp Var.Map.t] Term.visitor
-      method! enter_def t vars =
-        let v = Def.lhs t in
-        if Var.is_virtual v
-        then if Map.mem vars v
-          then Map.remove vars v
-          else match Def.rhs t with
-            | Bil.Unknown _ | Bil.Int _ as exp ->
-              Map.add vars ~key:v ~data:exp
-            | _ -> vars
-        else vars
+let is_alive dead t = not (Set.mem dead (Term.tid t))
+let live_phi dead blk = Term.filter phi_t ~f:(is_alive dead) blk
+
+let live_def checked dead blk =
+  Term.filter def_t blk ~f:(fun d ->
+      if checked (Def.lhs d) then is_alive dead d
+      else true)
+
+let substitute sub vars =
+  (object
+    inherit Term.mapper as super
+    method! map_var v =
+      match Map.find vars v with
+      | None -> Bil.var v
+      | Some e -> e
+    method! map_exp e = Exp.fold_consts (super#map_exp e)
+  end)#map_sub sub
+
+(* A simple constant propagation. Note, the input is required to be in SSA. *)
+let propagate_consts checked sub =
```

Comment:
this function does more work than needed, the term visitor will go through jmps and phi-nodes, maybe it is better to rewrite it as a fold over `Term.enum def_t blk`